components/raftstore/src/store/snap.rs

```diff
@@ -2099,18 +2091,28 @@ impl TabletSnapManager {
         PathBuf::from(&self.base).join(prefix)
     }
 
-    pub fn delete_snapshot(&self, key: &TabletSnapKey) -> bool {
-        let path = self.tablet_gen_path(key);
-        if path.exists() && let Err(e) = file_system::trash_dir_all(&path) {
-            error!(
-                "delete snapshot failed";
-                "path" => %path.display(),
-                "err" => ?e,
-            );
-            false
-        } else {
-            true
+    pub fn delete_snapshot(&self, key: TabletSnapKey) -> Result<()> {
+        for f in file_system::read_dir(&self.base)? {
```

Comment:
Why still need to iterate the directory?

Reply:
`new_without_init` only is only used as `SnapshotManagerCore` receive the v2 snapshot. And these test cases does not need to create the real dir, so it can pass. But if using `new` to create it, it will failed because the path is wrong.

Reply:
yes, it doesn't needs to iterate the directory. The backlogged snapshot task will be delete after the tikv restart.

Reply:
> new_without_init only is only used as SnapshotManagerCore receive the v2 snapshot. And these test cases does not need to create the real dir, so it can pass. But if using new to create it, it will failed because the path is wrong.

As explained before: If "" is meant to disable tablet snapshot, then tablet snapshot manager should be defined as an Option.

Reply:
OK, I will remove the constructed function in later pr.

Reply:
Why not just do it in this PR?

Reply:
done, have remove `init` and `new_without_init` functions .

Reply:
> > new_without_init only is only used as SnapshotManagerCore receive the v2 snapshot. And these test cases does not need to create the real dir, so it can pass. But if using new to create it, it will failed because the path is wrong.
> 
> As explained before: If "" is meant to disable tablet snapshot, then tablet snapshot manager should be defined as an Option.

Passing "" is not meant to disable tablet snap mgr. Some tests does not want to use snap manager but want to use struct that has the field of snap manager, so it pass "" to create the snap manager but never uses it. For snap manager, as it lazily create the dir, so it's fine. Now, we add tablet snap manager as a field of snap manager, creating dir based on "" may cause problem.

Reply:
I suggest to adjust these tests (very few), and add an assert that the path is not "" in the `SnapManagerBuilder::build`.

Reply:
> Some tests does not want to use snap manager but want to use struct that has the field of snap manager

What are those cases?

Reply:
I don't see `tablet_snap_manager` is used in the case. Can you show it in details?

Reply:
`TabletSnapManager` is a field of `SnapManager`
```
pub struct SnapManager {
    core: SnapManagerCore,
    max_total_size: Arc<AtomicU64>,

    // only used to receive snapshot from v2
    tablet_snap_manager: TabletSnapManager,
}
```

Reply:
So what's the problem when `tablet_snap_manager` is defined as `Option<TabletSnapManager>` and become `None` when it's not used?

Reply:
There's no problem for tablet_snap_manager to be defined as Option<TabletSnapManager>, but whether it is None or Some should not be determined by `path == ""` but by whether it's a tiflash engine.